New module for managing VMs and zones on SmartOS.

jasperla opened the pull request:
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (vmadm ecd4351f73) last updated 2017/01/04 18:36:55 (GMT +200)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

This is a new module that allows for creating and managing the state of zones
and virtual machines on SmartOS through `imgadm(1M)`. On SmartOS these VMs are
created using a JSON payload describing the VM.

This module supports most of the common options/properites that can be set for
a VM. Special care was taken to make it trivial to add support for new properties
with great ease (`properties` dict in `main()`). Not much validation is done on
the actual values as vmadm has proper error reporting and it would mean this
module would always have to keep up with imgadm in it's argument handling.

In addition to creating new VMs, this module can also manage the state. I.e. it
can reboot, shutdown, start and delete a single VM, or all VMs in a single task.

Two things are not supported yet:
- query for properties of an existing VM so they can be `register`-ed for use
  in other Ansible tasks.
- update properties of existing VMs. I was thinking of something along the lines
  of `state=updated`, however I first want to have a base to work on.

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->
```
  - name: Stop all zones
    vmadm:
      uuid: '*'
      state: stopped
```

results in:

```
TASK [Stop all zones] **********************************************************
ok: [calafate-gz] => {"changed": false, "state": "stopped", "uuid": "*"}
```

Or to create a new zone:
```
- name: create SmartOS zone
  vmadm:
    brand: joyent
    state: present
    alias: fw_zone
    image_uuid: 95f265b8-96b2-11e6-9597-972f3af4b6d5
    firewall_enabled: yes
    indestructible_zoneroot: yes
    nics:
      - nic_tag: admin
        ip: dhcp
        primary: true
    internal_metadata:
      root_pw: 'secret'
    quota: 1
```

Results in:
```
TASK [create SmartOS zone] *****************************************************
changed: [calafate-gz] => {"alias": "test1", "changed": true, "state": "created", "uuid": "421427b0-39ff-c934-c877-c119d0129ffe"}
```

@gundalow or @abadger would you please have a look?

jasperla committed:
New module for managing VMs and zones on SmartOS.

Currently allows for creating new instances and state changes.
Updating existing VMs is not supported yet

jasperla commented:
@bcoca  or @abadger, would you please review?

gundalow commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,560 @@
+    debug:
+        required: false
+        description:
+        - Don't delete the JSON payload used to generated a VM.
```
The indentation doesn't look right here (or on the other `descriptions`)

jasperla replied:
I've fixed that (and also for `imgadm`: https://github.com/ansible/ansible/pull/20148)

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,560 @@
+
+    module.exit_json(**result)
+
+from ansible.module_utils.basic import AnsibleModule
```
This should be on the top of the file.

jtyr commented:
After a quick look, these are my comments:

1) You are missing the RETURN = """ ... """ block.
2) The main() function seems to be tooooooooo long. You can improve it by creating functions per each state.
3) Use flake8 to improve your code.
4) Make sure it runs with Python 3.

jasperla committed:
Fix identation in DOCUMENTATION

jasperla committed:
Add RETURN block

jasperla committed:
Apply flake8 recommendations and split up main() into separate functions.

jasperla commented:
@jtyr Thanks, I've addressed points 1-3.
As for 4, the Shippable tests pass for Python 3 which seem good enough given that the default interpreter for SmartOS is still Python 2.

jasperla committed:
Be more about the states this module accepts and trim the exposed list

jasperla committed:
Be more verbose about states and the implications

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,600 @@
+def get_vm_prop(module, uuid, prop):
+    """Lookup a property for the given VM.
+    Returns the property, or None if not found."""
+    cmd = 'vmadm lookup -j -o {0} uuid={1}'.format(prop, uuid)
```
you should use the find_bin function to get vmadm and not rely on it being in PATH

bcoca replied:
since you use this multiple times I would either make it global to the script or attach it to the module object you are passing

jasperla replied:
vmadm is located on a read-only filesystem that cannot be moved to/from another location than `/usr/sbin/vmadm`. I will actually make `VMADM` globally available as `/usr/sbin/vmadm` since it's fixed and there is no other location it can reside.

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,600 @@
+    (rc, stdout, stderr) = module.run_command(cmd)
+
+    if rc != 0:
+        module.fail_json(msg=stderr)
```
a better message would not only include stderr but what you were attempting at the time (failed to get the properties associated with this vm: %stderr)

jasperla replied:
Good point, I've adjusted this elsewhere too.

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,600 @@
+            if not is_valid_uuid(vm_uuid):
+                module.fail_json(msg='Invalid UUID for VM {0}?'.format(vm_uuid))
+        else:
+            module.fail_json(msg='Could not retrieve UUID of newly created(?) VM')
```
so the tool can return rc == 0 but still fail?

jasperla replied:
I don't recall actually running into it, however a tool that prints it's normal output to `stderr` can be expected to do anything. So I'd rather err on the side of caution and handle such a case.

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,600 @@
+
+
+def create_payload(module, uuid):
+    """Create the JSON payload (vmdef) and return the filename."""
```
we normally prefer to use the same tmp directory as the modules themselves are put into

jasperla replied:
Is there a particular function to get that path or just `./` ?

jasperla replied:
As mentioned on IRC, I'll keep using `mkstemp()` for now, until there is a generic way of knowing where the Ansible temporary directory is. When that is implemented I will modify this module to store the payload file there and rely on `ANSIBLE_KEEP_REMOTE_FILES` to determine whether or not to keep the payload file).

bcoca replied:
finding it when not using pipeline is normally `dirname(__file__)` as module itself is located there.. but as discussed, need to create generic and then revisit modules doing this

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,600 @@
+version_added: "2.3"
+author: Jasper Lievisse Adriaanse (@jasperla)
+options:
+    debug:
```
modules already get a debug from how main ansible was called, you should not need to add the option here

jasperla replied:
I'll rename this appropriately to `save_payload`.

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,600 @@
+        module.fail_json(msg='Could not create valid JSON payload: {0}'.format(e))
+
+    # Create the temporary file that contains our payload, and set tight
+    # permissions for it may container sensitive information.
```
since it can contain sensitive info, is it possible to avoid writing to disk at all? piping it?

jasperla replied:
It is possible to feed it straight to `vmadm`, however the risk of someone snooping is small since there should be no other user in the global zone but root. I  do see your point and I'll add it to my list of things to add later. OK?

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,600 @@
+              ['image_uuid', module.params['image_uuid']]]:
+        if u[1] and u[1] != '*':
+            if not is_valid_uuid(u[1]):
+                module.fail_json(msg='{0} is not a valid UUID for {1}'.format(u[1], u[0]))
```
not a requirement, but would be nice to validate all IDs before returning an error (in case more than one is wrong)

jasperla replied:
That's indeed nice to have, rather than fixing one UUID and finding out another fails too. I'll fix that.

jasperla committed:
Fail with a more descriptive error messages than just 'stderr'

jasperla committed:
Valid all UUIDs before exiting

jasperla committed:
Incorporate feedback from bcoca

jasperla committed:
Print field name that contains invalid uuid, not the value

gundalow reviewed (approved):
Once bcoca is happy, I'm happy

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,612 @@
+version_added: "2.3"
+author: Jasper Lievisse Adriaanse (@jasperla)
+options:
+    save_payload:
```
i thought i had asked to remove this option in favor of keep remote files

jasperla replied:
Sorry, I thought that was about renaming the option. I'll zap it.

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,612 @@
+            shutdown the zone before removing it.
+            C(stopped) means the zone will be created if it doesn't exist already, before shutting
+            it down.
+    alias:
```
name is probably a better name for the name of the vm (alias can be an alias) .... had to parse that several times too 'meta'

jasperla replied:
Ok, I originally named it `alias` as that's the regular idiom. But I'll make alias an alias for name.

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,612 @@
+# generated JSON does not play well with the JSON parsers of Python.
+# The returned message contains '\n' as part of the stacktrace,
+# which breaks the parsers.
+VMADM = '/usr/sbin/vmadm'
```
i would use the find_bin function in modules, in case someone decides to move it in the future or run their custom version from a different path

jasperla replied:
Fair enough, I'll fix that.

bcoca commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,612 @@
+
+    if rc != 0:
+        module.fail_json(msg='Could not perform lookup of {0} on {1}: {2}'.
+                         format(prop, uuid, stderr))
```
not 2.4 compatible, checks are probably skipping everything in 'cloud' but considering the target platform, we might want to keep 2.4 compatibility in this case.

jasperla replied:
The target platform (SmartOS) does (only?) runs Python 2.7. As I mentioned here https://github.com/ansible/ansible/pull/19696#pullrequestreview-15343573 and was recommended here: https://github.com/ansible/ansible/pull/19696#pullrequestreview-15343221 I chose this particular syntax.

bcoca replied:
sorry, anything that smells like Sun makes me think they are still running Python2.2

jtyr replied:
Btw. the line wrapping is not PEP8 compliant. It should be like this:

```
module.fail_json(
    msg='Could not perform lookup of {0} on {1}: {2}'.format(prop, uuid, stderr))
```

Please use `flake8` to check your code before you push commits.

jasperla committed:
More changes as suggested by bcoca:

- drop save_payload completel
- make 'name' the options with an 'alias' alias
- use get_bin_path() for locating 'vmadm'

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+# along with Ansible. If not, see <http://www.gnu.org/licenses/>.
+#
+
+ANSIBLE_METADATA = {'status': ['preview'],
```
This could have prettier formatting:

```
ANSIBLE_METADATA = {
    'status': ['preview'],
    'supported_by': 'community',
    'version': '1.0'
}
```

jasperla replied:
There are hundreds other modules to adjust too. Won't it make more sense to do one big sweep?

gundalow replied:
No need to change what you've done, you can drop this issue.

The `ANSIBLE_METADATA` block you've got is consistent with the modules

jtyr replied:
Yes, ut's consistent with other ugly formatted modules. Even the author himself is using the right formatting on the line 357 (`cmds = {`) and 483 (`properties = {`) so why not to correct this as well?

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+module: vmadm
+short_description: Manage SmartOS virtual machines and zones
+description:
+    - Manage SmartOS virtual machines through vmadm(1M)
```
Standard indentation for YAML is two spaces. I can see you use 2 spaces in the `description` of the options and in `EXAMPLES` already. It would be nice to unify this across all the `DOCUMENTATION` and the `RETURN` block.

jtyr replied:
Please add dot at the end of the description.

jtyr replied:
I see you fixed the `DOCUMENTATION` block but did not fixed yet the `RETURN` block.

jtyr replied:
This needs to be addressed.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+    try:
+        return json.loads(stdout)[0]['uuid']
+    except:
+        return ''
```
I would prefer if the function would return `None` instead of empty string. Then do the appropriate test for `is None` on the return value of the function. The same applies to all functions in this module.

jasperla replied:
There are no other functions that return ''.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+        module.fail_json(msg='Could not retrieve UUID of {0}: {1}'.format(alias, stderr))
+
+    try:
+        return json.loads(stdout)[0]['uuid']
```
This looks a bit dirty. I would put only the `json.loads()` into the `try`/`exception` block and if the load was successful, I would test if if contains the `uuid`. The same applies to all places where you load JSON in this module.

jasperla replied:
This seemed like the most concise way of writing it. Otherwise it would be something like this, no?

```
    uuid = json_out[0]['uuid']
    if uuid:
        return uuid
    else:
        return None
```

jtyr replied:
Not really as there might not be any `[0]` and if there is there might not be any `uuid` in it. The right way of checking that is:

```
if len(json_output) > 0 and 'uuid' in json_out[0]:
    return
```

There is no need for `else:` as the default return value is `None` which is what we want if it fails.

jtyr replied:
This needs to be addressed.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+    (rc, stdout, stderr) = module.run_command(cmd)
+
+    if rc != 0:
+        module.fail_json(msg='Could not retrieve UUID of {0}: {1}'.format(alias, stderr))
```
The `stderr` should be used in the `exception=stderr` keyword. The same applies to all places where you use `stderr` in this module.

jasperla replied:
I wasn't aware of this practice, but sure, I'll adjust that.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+    # Create the temporary file that contains our payload, and set tight
+    # permissions for it may container sensitive information.
+    try:
+        # XXX: When there's a way to get the current ansible temporary directory
```
What does the `XXX` mean here?

jasperla replied:
Well, as mentioned by bcoca, there should be a generic function that will tell us the name of the temporary directory created by Ansible. Once we have said function it can be used here.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+
+
+def manage_all_vms(module, vm_state, changed):
+    """ Handle operations for all VMs, which can by definition only
```
I would either use everywhere the simple comments ('# ...`) or use everywhere the comment blocks (`""" ... """`) but I wouldn't combine both.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+        # Shortcut for check mode, if there is no VM yet, it will need to be created.
+        # Or, if the VM is not in the desired state yet, it needs to transition.
+        if (not current_vm_state) or (get_vm_prop(module, uuid, 'state') != state):
+            result['changed'] = True
```
Either use `result['changed'] = ` and remove the `changed =` OR use `changed =`  and remove the `result['changed']` but do not use both. That applies to all `main()` function.

jasperla replied:
Good point, I'll adjust that accordingly.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+version_added: "2.3"
+author: Jasper Lievisse Adriaanse (@jasperla)
+options:
+    name:
```
It would be nice to sort the options alphabetically.

jasperla replied:
Done

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+DOCUMENTATION = '''
+---
+module: vmadm
+short_description: Manage SmartOS virtual machines and zones
```
Please add dot at the end of the description.

jasperla replied:
I'll fix that.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+        required: false
+        description:
+          - Resolvers in C(/etc/resolv.conf) will be updated when updating
+            the C(resolvers) property.
```
You are definitely confusing markup for options and the values. The `resolvers` is an option so it should be wrapped in `I(..)` in the documentation. The `/etc/resolv.conf` is a value so it should be wrapped in `C(..)` in the documentation. Please go through all the documentation block and fix it accordingly.

jasperla replied:
This was the only offender AFAICS that referred to an option as `C()` instead of `I()`.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+        description:
+          - ZFS pool the VM's zone dataset will be created in.
+requirements:
+    - python >= 2.6
```
I would use `Python` instead of the `python`.

jasperla replied:
There is very little precedent for that. I think it should be `python` as that is what most package managers call it too.

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -0,0 +1,609 @@
+    sample: 'running'
+'''
+
+from ansible.module_utils.basic import AnsibleModule
```
Shouldn't the imports be above any variable definitions?

jasperla replied:
It's all over the place in other modules too, but sure, I'll move it.

jasperla committed:
Doc fixes

Change-Id: Ia23704c99d24b7782d8884fd22e11af61cef724c

jasperla committed:
More fixes suggested by jtyr

Change-Id: Ia3e65277bf5a499651c70b096e2988d7cc9eceb5

jasperla committed:
Drop local changed variable

Change-Id: Ia8beaa615ba4a47004d7af791c28a5a5855bcf61

jasperla committed:
Use correct ANSIBLE_METADATA format

Change-Id: I51ded34e543005933c7c739501388e24e5f80c1d

jasperla committed:
Re-indent RETURN

Change-Id: I0c266d7a83517be54fc6242dfb0496a0a480f628

jasperla committed:
Fix documentation formatting for options and fix an unused variable

jasperla committed:
Move imports to their original location again

jasperla committed:
Fail if there's invalid json returned instead of trying to continue

jtyr commented on lib/ansible/modules/cloud/smartos/vmadm.py:
```diff
@@ -277,10 +277,19 @@ def get_vm_uuid(module, alias):
+      try:
+        stdout_json = json.loads(stdout)
+      except:
+        module.fail_json(msg='Invalid JSON returned by vmadm for uuid lookup of {0}'.format(alias))
```
I would always provide more information why that happened:

```
from ansible.module_utils.pycompat24 import get_exception

e = get_exception()
module.fail_json(
    msg='Invalid JSON returned by vmadm for uuid lookup of {0}'.format(alias),
    exception=str(e))
```

abadger replied:
If getting the exception message, should put it in the message and use ```to_native(e)``` to retrieve it.  (to_native will avoid tracebacks if the string representation of e contains non-ascii).

if you want to use the exception field, it needs to contain the traceback which can be retrieved via the stdlib function: ```traceback.format_exc()```.

jasperla committed:
Betterer error handling for json.loads()

abadger commented:
Merged to devel
